src/main/java/com/teragrep/pth_06/planner/offset/HdfsOffset.java

```diff
@@ -0,0 +1,109 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ *
+ * Additional permission under GNU Affero General Public License version 3
+ * section 7
+ *
+ * If you modify this Program, or any covered work, by linking or combining it
+ * with other code, such other code is not for that reason alone subject to any
+ * of the requirements of the GNU Affero GPL version 3 as long as this Program
+ * is the same Program as licensed from Suomen Kanuuna Oy without any additional
+ * modifications.
+ *
+ * Supplemented terms under GNU Affero General Public License version 3
+ * section 7
+ *
+ * Origin of the software must be attributed to Suomen Kanuuna Oy. Any modified
+ * versions must be marked as "Modified version of" The Program.
+ *
+ * Names of the licensors and authors may not be used for publicity purposes.
+ *
+ * No rights are granted for use of trade names, trademarks, or service marks
+ * which are in The Program if any.
+ *
+ * Licensee must indemnify licensors and authors for any liability that these
+ * contractual assumptions impose on licensors and authors.
+ *
+ * To the extent this program is licensed as part of the Commercial versions of
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.pth_06.planner.offset;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.spark.sql.connector.read.streaming.Offset;
+
+import java.io.Serializable;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+
+// Class for representing a serializable offset of HDFS data source.
+// S3 has epoch hours as offsets, kafka has native TopicPartition offsets and HDFS should have file-metadata (use same format as in Kafka, topicpartition + record offset, which can be extracted from the metadata).
+
+public class HdfsOffset extends Offset implements Serializable {
+
+    // TODO: Implement everything that is needed for tracking the offsets for HDFS datasource.
+
+    private static final Type mapType = new TypeToken<Map<String, Long>>() {
+    }.getType();
+    private final Map<String, Long> serializedHdfsOffset;
+
+    public HdfsOffset(Map<TopicPartition, Long> offset) {
+        serializedHdfsOffset = new HashMap<>(offset.size());
+        for (Map.Entry<TopicPartition, Long> entry : offset.entrySet()) {
```

Comment:
could this be removed from constructor?

Reply:
Map<TopicPartition, Long> is not serializable which is why it is converted to serializable Map<String, Long> in the constructor. Changed the constructor input parameter to Map<String, Long>, conversion to serializable map is now done before the object is initialized.
Solved in commit 48062e0798c8b43a1d3dc543b5d718392ad71775